diff -up dhcp-4.1.1-P1/client/dhclient.c.logpid dhcp-4.1.1-P1/client/dhclient.c
--- dhcp-4.1.1-P1/client/dhclient.c.logpid	2010-11-16 17:20:31.000000000 +0100
+++ dhcp-4.1.1-P1/client/dhclient.c	2010-11-16 17:20:31.000000000 +0100
@@ -153,7 +153,7 @@ main(int argc, char **argv) {
 	else if (fd != -1)
 		close(fd);
 
-	openlog("dhclient", LOG_NDELAY, LOG_DAEMON);
+	openlog("dhclient", LOG_NDELAY | LOG_PID, LOG_DAEMON);
 
 #if !(defined(DEBUG) || defined(__CYGWIN32__))
 	setlogmask(LOG_UPTO(LOG_INFO));
